### ISSUE TITLE
Add basic websocket support for GDAX

### ISSUE THREAD

**Issue by DeviaVir** [highvelocityspace/zenbot#54]
_From @krystophv on December 21, 2017 22:24_

This PR makes a first step into reading data from the websocket stream for GDAX instead of continually polling their REST APIs. It doesn't feed the websocket reports directly into the system, but rather sets up a couple of objects/arrays that will act as a buffer or cache for the incoming websocket data and then when the `exchange` level API functions run, they're set to check the websocket cache (or websocket fed live orderbook) before making a REST call. This should initially alleviate REST calls from `getTrades` and `getQuote`. At a minimum, should also be able to write a little more processing code and set up a similar cache for `getOrder` in the near future.

_Copied from original issue: carlos8f/zenbot/pull/970_

**Comment by DeviaVir** [comment]
_From @fuzzyTew on December 26, 2017 11:11_

This is great!  It might help people review the changes if they were squashed and rebased onto master with only relevent commit names retained (requires a git push --force to the branch to change the history).

**Comment by DeviaVir** [comment]
_From @Fredrik81 on December 27, 2017 10:11_

I got an issue testing this change. I could reproduce it like this:
1. git pull latest zenbot
2. pull this PR
3. create/copy in conf.js with API keys for gdax
4. try to backfill (i tried gdax.ETH-BTC)

It will fail with message that not all keys are filled in. Nothing wrong with the the conf-file it works in live trading.
If i remove the conf.js it works to backfill.

**Comment by DeviaVir** [comment]
_From @fuzzyTew on December 27, 2017 12:39_

@Fredrik81, I thought that issue was fixed in 22c569e 17 hours ago.  When did you perform this test?

Backfill works fine for me.  I haven't verified the data for correctness, though.

**Comment by DeviaVir** [comment]
_From @fuzzyTew on December 27, 2017 12:55_

I was reading the gdax docs and found https://docs.gdax.com/#sequence-numbers stating that some messages will be dropped or sent out of order.  They say at https://docs.gdax.com/#the-code-classprettyprintmatchescode-channel that the heartbeat channel should be used to identify and resync when information is lost, using either sequence numbers or trade ids.  I don't know if this is a huge issue if a trade is dropped here and there, but it's perhaps the next thing missing.

**Comment by DeviaVir** [comment]
_From @luponata on December 27, 2017 13:8_

With same steps of Fredrik81 i get:

(node:4844) DeprecationWarning: `new PublicClient()` no longer accepts a product ID as the first argument. 
(node:4844) DeprecationWarning: `getProductTrades()` now requires a product ID as the first argument. Attempting to use PublicClient#productID (LTC-EUR) instead

while backfilling

same for --manual on live trade

tested now, with corrected exchange.json of 22c569e08a14ec2657daadf572296e19595ad1b2

**Comment by DeviaVir** [comment]
_From @Fredrik81 on December 27, 2017 13:52_

Hello,
Tested same steps again using a new git pull so all the latest merged PR's included.
Give same result as @luponata.
I hope your able to fix it :-)

zenbot$ ./zenbot.sh backfill gdax.ETH-BTC --days=2
node-telegram-bot-api deprecated Automatic enabling of cancellation of promises is deprecated.
In the future, you will have to enable it yourself.
See https://github.com/yagop/node-telegram-bot-api/issues/319. module.js:635:30
(node:6907) DeprecationWarning: `new PublicClient()` no longer accepts a product ID as the first argument. 
(node:6907) DeprecationWarning: `getProductTrades()` now requires a product ID as the first argument. Attempting to use PublicClient#productID (ETH-BTC) instead.

gdax.ETH-BTC saved 100 trades 2 days left
(node:6907) DeprecationWarning: `getProductTrades()` now requires a product ID as the first argument. Attempting to use PublicClient#productID (ETH-BTC) instead.
.(node:6907) DeprecationWarning: `getProductTrades()` now requires a product ID as the first argument. Attempting to use PublicClient#productID (ETH-BTC) instead.
websocket connection to ETH-BTC opened
.(node:6907) DeprecationWarning: `getProductTrades()` now requires a product ID as the first argument. Attempting to use PublicClient#productID (ETH-BTC) instead.
.(node:6907) DeprecationWarning: `getProductTrades()` now requires a product ID as the first argument. Attempting to use PublicClient#productID (ETH-BTC) instead.
.(node:6907) DeprecationWarning: `getProductTrades()` now requires a product ID as the first argument. Attempting to use PublicClient#productID (ETH-BTC) instead.

**Comment by DeviaVir** [comment]
_From @luponata on December 27, 2017 14:16_

Anyway i don't know if these tests should be done after the successful merge of https://github.com/coinbase/gdax-node/pull/179

**Comment by DeviaVir** [comment]
_From @fuzzyTew on December 27, 2017 15:12_

@luponata @Fredrik81 I don't get those warnings.  This PR updates the version of gdax-node depended on in package.json -- have you run npm install to download that updated version?

**Comment by DeviaVir** [comment]
_From @luponata on December 27, 2017 15:13_

yes i did :/

**Comment by DeviaVir** [comment]
_From @fuzzyTew on December 27, 2017 15:20_

Looks like they changed the API last night with https://github.com/coinbase/gdax-node/pull/178

**Comment by DeviaVir** [comment]
_From @fuzzyTew on December 27, 2017 18:37_

Pinned the dependency to the old API for now with https://github.com/krystophv/zenbot/pull/2

**Comment by DeviaVir** [comment]
_From @krystophv on December 27, 2017 18:59_

Thanks everyone for the interest and testing. I merged @fuzzyTew's PR pinning the dependency to a tag so the dependency doesn't update with breaking changes unexpectedly in the future. I guess we'll just need to check up on the dependency periodically to see if it's got a new tag (and at that point update it). Hopefully they just start using NPM properly, but that might be a bit much to hope for :/

**Comment by DeviaVir** [comment]
_From @fuzzyTew on December 27, 2017 19:26_

I've tried to update the code for the development changes in gdax-node at https://github.com/fuzzyTew/zenbot/tree/gdax_websocket_subscribe mostly so I can use the subscribe/unsubscribe change, but it also has changes equivalent to @Fredrik81's now.  I haven't really tested it.

**Comment by DeviaVir** [comment]
_From @Fredrik81 on December 27, 2017 21:2_

It updates/download data super quick. this will be a big improvement when live.

**Comment by DeviaVir** [comment]
_From @krystophv on December 27, 2017 21:13_

Once this lands, I'll have another PR up shortly that will build out the local orders cache based on the websocket feed so we can avoid making getOrder calls too.

**Comment by DeviaVir** [comment]
_From @Fredrik81 on December 27, 2017 21:30_

I can confirm that with the last API version change the problem i faced is gone.

**Comment by DeviaVir** [comment]
_From @luponata on December 28, 2017 0:49_

solved

**Comment by DeviaVir** [comment]
_From @krystophv on December 28, 2017 1:22_

@luponata that's clearly a rate limit exceeded error. and not entirely relevant to this PR discussion. if you feel it's an issue with fuzzy's changes, then maybe making an issue on their branch would be appropriate, or perhaps asking around in the Discord. I'd recommend turning down the poll rate.

**Comment by DeviaVir** [comment]
_From @luponata on December 28, 2017 1:29_

my fault, was using a copy of conf-sample.js from another repository with too low ms on poll-trades..
instead of copying every times my api keys

**Comment by DeviaVir** [comment]
This issue was moved to highvelocityspace/zenbot#24
